pkgtools/binpatch/files/binpatch.1
.\"	$NetBSD: binpatch.1,v 1.2 2013/07/20 21:50:52 wiz Exp $
.\"
.\" Copyright (c) 2004 by Andrew Brown <atatat@netbsd.org>
.\" Absolutely no warranty.
.\"
.Dd July 20, 2004
.Dt BINPATCH 1
.Sh NAME
.Nm binpatch
.Nd trivial binary patch applicator
.Sh SYNOPSIS
.Nm
.Pa file=...
.Pa size=...
.Pa offset=...
.Pa compare=...
.Pa skip=...
.Pa replace=...
.Sh DESCRIPTION
The
.Nm
utility can read and replace a small section of a given file.
It is designed for use in those instances where a problem exists with
a given binary that cannot be reconstructed from source code, but the
required change can be implemented by replacing a few bytes in the
existing binary.
All arguments must be given.
.Sh EXAMPLES
Given a binary called
.Dq a.out
of 10713 bytes in size with the following text segment:
.Bd -literal -offset indent
% objdump -h a.out
.sp
a.out:     file format elf32-i386
.sp
Sections:
Idx Name          Size      VMA       LMA       File off  Algn
[...]
  9 .text         00000be4  08048968  08048968  00000968  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
[...]
% objdump -d -j .text a.out
[...]
 8048b0f:       83 ef 04                sub    $0x4,%edi
 8048b12:       ff d0                   call   *%eax
 8048b14:       83 fe ff                cmp    $0xffffffff,%esi
[...]
.sp
.Ed
where we wish to elide the call through
.Ar %eax
by replacing it with a series of
.Ar nop
(or
.Dq no operation )
instructions (the machine code for this on the i386 platform is 0x90),
we first calculate the offset into the file of the previous
.Ar sub
instruction.  To do this, we take the address of the
.Ar sub
instruction as given by the dissassembly output, subtract the
.Dq LMA
and add the
.Dq File off
values from the objdump output (note that
.Xr bc 1
expects hexadecimal values to be given using upper case):
.Bd -literal -offset indent
% bc
ibase=16
8048B0F-08048968+00000968
2831
.sp
.Ed
The region of the binary we want to compare to before applying the
patch is the concatenation of the relevant machine codes from the
dissassembly dump (\c
.Ar 83ef04ffd083feff )
and the replacement is simply two
.Ar nop
instructions (\c
.Ar 9090 ) ,
that will replace the
.Ar ffd0
of the original call.
The offset of the replacement is 3, since that is the number of bytes
in the
.Ar sub
instruction.
From this we have our patch:
.Bd -literal -offset indent
% binpatch file=a.out size=10713 offset=2831 \\
  compare=83ef04ffd083feff skip=3 replace=9090
% objdump -d -j .text a.out
[...]
 8048b0f:       83 ef 04                sub    $0x4,%edi
 8048b12:       90                      nop    
 8048b13:       90                      nop    
 8048b14:       83 fe ff                cmp    $0xffffffff,%esi
[...]
.sp
.Ed
And thus the call is removed.
.Sh DIAGNOSTICS
The diagnostics are terse and almost unhelpful, but are more verbose
than users of
.Xr ed 1
might be used to.
They typically mention the command line argument that was in error.
.Sh SEE ALSO
.Xr bc 1 ,
.Xr objdump 1 ,
.Xr patch 1
.Sh AUTHORS
.An Andrew Brown Aq Mt atatat@netbsd.org


doc/NEWS
# $NetBSD: NEWS,v 1.33 2025/12/03 21:57:26 leot Exp $

PKGSRC NEWS
===========

  This file contains information about important infrastructural changes in
  pkgsrc. The intended readers are the pkgsrc developers as well as anyone
  tracking pkgsrc-current.

m68k softfloat (m68ksf) support [nat 2025-10-22]

        Support for m68k softfloat was added to pkgsrc.

Default versions bumped [wiz 2025-10-06]

	PYTHON_VERSION_DEFAULT is now 313.

Default MySQL implementation changed to MariaDB 11.4 [nia 2024-12-25]

	The default version of MySQL as set by the pkgsrc infrastructure
	was changed from MariaDB 10.11 to MariaDB 11.4.

Default versions bumped [wiz 2024-08-12]

	PYTHON_VERSION_DEFAULT is now 312.

Default versions bumped [taca 2024-02-10]

	RUBY_VERSION_DEFAULT is now 32.

Default MySQL implementation changed to MariaDB 10.11 [nia 2024-01-23]

	The default version of MySQL as set by the pkgsrc infrastructure
	was changed from MariaDB 10.6 to MariaDB 10.11.

Default versions bumped [wiz 2023-08-14]

	PYTHON_VERSION_DEFAULT is now 311.

Support for USE_CC_FEATURES and USE_CXX_FEAUTRES [nia 2023-07-18]

	USE_CC_FEATURES and USE_CXX_FEATURES can now be used
	to specify a minimum level of C and C++ support required
	to compile a package, including specific libstdc++ features.

	ALLOW_NEWER_COMPILER can be set to no in mk.conf to skip
	building packages that require a newer compiler than the system
	provides.

Default versions bumped [nia 2022-06-30]

	PYTHON_VERSION_DEFAULT is now 310.
	PGSQL_VERSION_DEFAULT is now 14.
	LUA_VERSION_DEFAULT is now 54.

Dropped support for compiling glib2-using packages without introspection
support [wiz 2022-06-07]

PKGSRC_USE_MKTOOLS support added [jperkin 2022-06-07]

        Added C-based tools to speed up various parts of the
        mk infrastructure via PKGSRC_USE_MKTOOLS=yes in mk.conf.

        For packages with large numbers of dependencies, you should see
        anywhere up to around 10x speed-up for the 'wrapper' phase, and
        perhaps more importantly a huge reduction in system time usage.

Linux infrastructure adapted to non-standard distributions [nia 2022-05-01]

        The Linux pkgsrc infrastructure was adapted to support Linux
        distributions where binaries and shared objects are not
        stored in a normal Unix-like file system layout.  It's hoped
        that this will be useful for pkgsrc development on platforms
        like NixOS and Gobo Linux.

UnixWare support [nia 2021-12-26]

        Support for SCO/Xinuos UnixWare was improved to the point where
        it is possible to complete a bootstrap on this platform.

SHA1 distfile hashes deprecated [nia 2021-10-07]

        Distfile hash algorithms have been changed, keeping SHA512,
        adding BLAKE2s, and removing SHA1 and RMD160.  New packages
        and updates will only generate SHA512 and BLAKE2s, and the
        stored hashes have been adjusted.

Default Python version changed to 3.9 [nia 2021-10-01]

Increased enabled-by-default hardening options [nia 2021-09-27]

	PKGSRC_USE_MKPIE was enabled on some architectures.
	Packages that provide static libraries to other packages
	should be rebuilt with MKPIE enabled.
	Some packages may experience runtime errors on platforms
	like i386 - usually this is due to unsafe assembly code,
	which should be disabled when MKPIE is enabled.

	PKGSRC_USE_SSP was bumped from "yes" to "strong".  This
	is not expected to have a noticable effect.

	More information:
	https://www.NetBSD.org/docs/pkgsrc/hardening.html
	(or see "Security hardening" section of doc/pkgsrc.txt)

Default MySQL implementation changed to MariaDB 10.6 [nia 2021-09-27]

Default PostgreSQL changed to 13 [nia 2021-06-28]

Default MySQL implementation changed to MariaDB [nia 2021-06-24]

	The default version of MySQL as set by the pkgsrc infrastructure
	was changed from MySQL 5.7 to MariaDB 10.5.

	This is configurable by setting MYSQL_VERSION_DEFAULT in mk.conf.

	PHP packages are no longer tied to a specific MySQL version
	due to the native driver being enabled by default.

Python 3.8 default [nia 2020-12-04]
	The default Python version was switched to 3.8.

Default database location changed for NetBSD [wiz 2020-12-02]
	The database of installed packages now defaults to ${PREFIX}/pkgdb
	instead of /var/db/pkg on NetBSD. This matches the default on
	other operating systems and cases where a non-default bootstrap
	is used.

	To deal with this change, see this blog post:

	https://pkgsrc.org/news/pkgdb-change/

	Since this change affects the packaging tools only, there is no change
	to binary packages and they do not need to be rebuilt.

lua 5.3 default [nia 2020-03-30]
	The default Lua version was switched to 5.3.
	Since many packages specify a preferred Lua version, this is mainly
	expected to affect users of Lua modules who were using the default
	settings.

gfortran default [nia 2020-03-30]
	gfortran is now the default Fortran compiler instead of g95.
	gfortran is expected to support more pkgsrc platforms and be
	better at compiling newer Fortran code.

python 3.7 default [maya 2019-04-24]
	As Python 2.x versions are nearing end of life, we have switched
	the default to a newer version.
	Python 2.7 packages are still available, and most python packages
	are built for all available python versions.

Lua multiversion support [alnsn 2014-05-03]
	Lua packages lang/lua51 and lang/lua52 have been modified
	to coexist without conflicts, infrastructure for supporting
	multiple versions of Lua has been put in place and all
	dependent packages have been adapted to the new infrastructure.

SMF support is introduced [jperkin 2014-03-11]

Unprivileged staged installation is default [asau 2012-08-14]
	pkgsrc has changed to use unprivileged staged installation by default.
	This makes it possible to prevent some installation mistakes.
	It also makes detection of package conflicts easier.
	Previously you had to set PKG_DESTDIR_SUPPORT to "user-destdir"

pkgtools/binpatch/files/binpatch.c
/*      $NetBSD: binpatch.c,v 1.1.1.1 2004/07/23 03:37:26 atatat Exp $ */

/*
 * ------------------------------------------------------------------------
 * "THE BEER-WARE LICENSE" (Revision 42):
 * Andrew Brown <atatat@NetBSD.org> wrote this file.  As long as you
 * retain this notice you can do whatever you want with this stuff.
 * If we meet some day, and you think this stuff is worth it, you can
 * buy me a beer in return.
 * ------------------------------------------------------------------------
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int
die(int rc, const char *msg)
{
	if (rc)
		perror(msg);
	else
		fprintf(stderr, "%s\n", msg);
	exit(1);
}

static void
cvt(const char *s, const char *i, unsigned char *o, size_t l)
{
	int t, x;

	for (t = 0; t < l; t++) {
		x = i[2 * t];
		if (x >= '0' && x <= '9')
			x -= '0';
		else if (x >= 'a' && x <= 'f')
			x -= 'a' - 10;
		else if (x >= 'A' && x <= 'F')
			x -= 'A' - 10;
		else
			die(0, s);
		o[t] = x * 16;

		x = i[2 * t + 1];
		if (x >= '0' && x <= '9')
			x -= '0';
		else if (x >= 'a' && x <= 'f')
			x -= 'a' - 10;
		else if (x >= 'A' && x <= 'F')
			x -= 'A' - 10;
		else
			die(0, s);
		o[t] += x;
	}
}

int
main(int argc, char *argv[])
{
	struct stat st;
	char *key, *value;
	int f, i;
	unsigned char *buf, *get, *put;
	size_t lget, lput;
	off_t sz, cmp_off, skip_off;

	f = -1;
	get = put = NULL;
	lget = lput = 0;
	sz = cmp_off = skip_off = -1;

	while (--argc > 0) {
		key = *++argv;
		if ((value = strchr(key, '=')) == NULL)
			die(0, "value required");
		else
			*value++ = '\0';

		if (strcmp(key, "file") == 0) {
			f = open(value, O_RDWR);
			if (f == -1)
				die(1, "file");
		}
		else if (strcmp(key, "size") == 0) {
			char *t;
			errno = 0;
			sz = strtol(value, &t, 0);
			if (errno != 0)
				die(1, "size");
		}
		else if (strcmp(key, "offset") == 0) {
			char *t;
			errno = 0;
			cmp_off = strtol(value, &t, 0);
			if (errno != 0)
				die(1, "offset");
		}
		else if (strcmp(key, "compare") == 0) {
			lget = strlen(value);
			if (lget % 2 != 0)
				die(0, "compare");
			lget /= 2;
			get = malloc(lget);
			buf = malloc(lget);
			cvt("compare", value, get, lget);
		}
		else if (strcmp(key, "skip") == 0) {
			char *t;
			errno = 0;
			skip_off = strtol(value, &t, 0);
			if (errno != 0)
				die(1, "offset");
		}
		else if (strcmp(key, "replace") == 0) {
			lput = strlen(value);
			if (lput % 2 != 0)
				die(0, "replace");
			lput /= 2;
			put = malloc(lput);
			cvt("replace", value, put, lput);
		}
	}

	/*
	 * ./binpatch
	 *    file=${MOZILLA_HOME}/netscape
	 *    size=13823336
	 *    offset=0x008073e9
	 *    compare=6a00e82406a3ffe81f0ca3ff
	 *    skip=2
	 *    replace=9090909090
	 */

	if (f == -1)
		die(0, "file missing");
	if (get == NULL || lget == 0)
		die(0, "compare missing");
	if (put == NULL || lput == 0)
		die(0, "replace missing");
	if (sz == -1)
		die(0, "size missing");
	if (cmp_off == -1)
		die(0, "offset missing");
	if (skip_off == -1)
		die(0, "skip missing");
	if (skip_off < 0 ||
	    (skip_off == 0 && lput >= lget) ||
	    (skip_off > 0 && skip_off + lput > lget))
		die(0, "illegal skip");

	if (fstat(f, &st) == -1)
		die(1, "fstat");
	if (st.st_size != sz)
		die(0, "wrong size");
	if (lseek(f, cmp_off, SEEK_SET) == -1)
		die(1, "lseek");
	if (read(f, buf, lget) != lget)
		die(1, "read");
	if (memcmp(buf, get, lget) != 0)
		die(0, "instructions not found");
	if (lseek(f, cmp_off + skip_off, SEEK_SET) == -1)
		die(1, "lseek");
	if (write(f, put, lput) != lput)
		die(1, "write");
	if (close(f) != 0)
		die(1, "close");

	return (0);
}
